[quant][graphmode][refactor] split getInvokedMethods

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30553 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #30552 [quant][graphmode][refactor] Add registerQParams function
* #30551 [quant][graphmode] Insert GetAttr for quantization parameters instead of Constant
* #30550 [quant][graphmode] Make `InsertQuantDeQuantHelper` global
* #30549 [quant][graphmode] Rename QuantizeHelper to InsertQuantDeQuantHelper
* #30548 [quant][graphmode] InsertObservers for shared class types
* #30547 [quant][graphmode][refactor] Globally record observer nodes
* **#30546 [quant][graphmode][refactor] split getInvokedMethods**

Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18776304](https://our.internmc.facebook.com/intern/diff/D18776304)

jerryzh168 committed:
[quant][graphmode][refactor] insertObserversForInvokedMethods

Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM reviewed (changes requested):
I agree that this part of the code can be refactored, but I'm not sure about the current version. Currently `insertObservers` calls `insertObserversForInvokedMethods` - for a reader it's not clear what the difference between these two is. It is not clear what makes a method "invoked" and why we pass a method name as an argument there. In general, functions should ideally lower the abstraction level as they call each other. For instance, when `insertObservers` calls `addIntermediateValuesToSkipObserver` it makes sense - the latter is logically an inner step of the former, and it is not the case for `insertObservers` and `insertObserversForInvokedMethods`.

I think we should use a better name for `insertObserversForInvokedMethods` and it should clearly convey its semantics. I think it actually might be better to keep the function in the loop - in this case it would be easier to find a descriptive name for what it does.

jerryzh168 commented:
> I agree that this part of the code can be refactored, but I'm not sure about the current version. Currently `insertObservers` calls `insertObserversForInvokedMethods` - for a reader it's not clear what the difference between these two is. It is not clear what makes a method "invoked" and why we pass a method name as an argument there. In general, functions should ideally lower the abstraction level as they call each other. For instance, when `insertObservers` calls `addIntermediateValuesToSkipObserver` it makes sense - the latter is logically an inner step of the former, and it is not the case for `insertObservers` and `insertObserversForInvokedMethods`.
> 
> I think we should use a better name for `insertObserversForInvokedMethods` and it should clearly convey its semantics. I think it actually might be better to keep the function in the loop - in this case it would be easier to find a descriptive name for what it does.

I think we'll have a new insertObservers function in the future which calls `insertObserversForInvokedMethods` and a *new* `insertObserversForGraph`(name subject to change) which is the current implementation of `insertObservers`
This refactoring is needed for shared types support so we can't keep this inside the loop

ZolotukhinM commented:
> This refactoring is needed for shared types support so we can't keep this inside the loop

This is fine, we just need to come up with a better name then (maybe for the outer function too). But every PR should leave the code in a better (or at least not worse) state, while this change I think makes the code harder to understand.

jerryzh168 commented:
> This is fine, we just need to come up with a better name then (maybe for the outer function too). But every PR should leave the code in a better (or at least not worse) state, while this change I think makes the code harder to understand.

sounds good, I think we just need to have a better name for outer function that covers both things we do, e.g. `insertObserversForMethodAndInvokedMethods`

jerryzh168 committed:
Update on "[quant][graphmode][refactor] insertObserversForInvokedMethods"


Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] insertObserversForInvokedMethods"


Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] insertObserversForInvokedMethods"


Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] split getInvokedMethods"


Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -147,6 +148,10 @@ class InsertObserversHelper {
   void insertObservers(script::Module& module, const std::string& method_name);
 
  private:
+  ModuleMethodVector getInvokedMethods(
```
The signature would be simpler if we just returned `std::vector<Method*>` or `std::vector<Graph*>` (`insertObservers` can also accept method/graph instead of module-method_name pair). Could you please change that?

ZolotukhinM reviewed (approved):
Looks good! I have only one remark - pls see inline.

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -271,6 +276,50 @@ graph(%a, %w, %b, %stride, %padding, %dilation, %transposed, %output_padding, %g
+        script::Module callee_module;
+        if (module_instance->node()->kind() == prim::GetAttr) {
+          auto child_module_name = module_instance->node()->s(attr::name);
+          callee_module = module.attr(child_module_name).toModule();
```
@ZolotukhinM we are using `module` here, so we'll need to keep the signature..

jerryzh168 committed:
Update on "[quant][graphmode][refactor] split getInvokedMethods"


Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18776304](https://our.internmc.facebook.com/intern/diff/D18776304)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][refactor] split getInvokedMethods"


Summary:
factor out this function for later support of quantizing shared types

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18776304](https://our.internmc.facebook.com/intern/diff/D18776304)

[ghstack-poisoned]
